### PR TITLE
Fix ad images not always loading due to race condition

### PR DESCRIPTION
**Pull request by dwick**
:eyeglasses: @curioussavage 

Can you test this on staging please?

### CODE REVIEW

**Comment by dwick** [comment]
:eyeglasses: @ajacksified for `ListingList` changes

File: src/views/components/ListingList.jsx [hunk 1]
```diff
@@ -33,28 +33,34 @@ class ListingList extends React.Component {
     this.props.app.off(constants.RESIZE, this._resize);
   }
 
+  _getWindowHeight () {
+    return window.innerHeight * 2;
```
Comment (ajacksified) [inline src/views/components/ListingList.jsx]: I like moving it to a function, but calling it `windowheight` and returning `windowheight*2` doesn't align.

File: src/views/components/ListingList.jsx [hunk 2]
```diff
@@ -124,14 +135,13 @@ class ListingList extends React.Component {
...
 
     // If ads are enabled, splice an ad into the listings.
-    if (this.props.showAds && listings.length) {
-      ad = this.buildAd();
-      listings.splice(this.state.adLocation, 0, ad);
+    if (props.showAds && listings.length) {
+      listings.splice(this.state.adLocation, 0, this.buildAd());
     }
 
-    this.length = listings.length;
```
Comment (danehansen) [inline src/views/components/ListingList.jsx]: i created an instance variable of length here so that length doesn't have to be recalculated every iteration of the for loop each time _scroll is called

Comment (dwick) [inline src/views/components/ListingList.jsx]: http://stackoverflow.com/a/5752991/704286

Comment (danehansen) [inline src/views/components/ListingList.jsx]: that is good news!

**Comment by curioussavage** [comment]
I can push it up to staging but so far locally I'm not seeing ad images

**Comment by curioussavage** [comment]
thumbnail is getting deleted when you get a response and build the link but in `componentDidMount` in Ad.jsx when you turn the json version back into a link instance there is no `thumbnail` or `mobile_ad_url` prop so it ends up with the default img

**Comment by curioussavage** [comment]
:nail_care:

**Comment by curioussavage** [comment]
is there a reason why we are converting models to json before returning them and the just instantiating them again? I have wondered that while looking at other things. @ajacksified ?

**Comment by ajacksified** [comment]
@curioussavage when snoode was first created, it assumed we could use [proxies](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy); since it can't, we turn it into JSON before returning them, otherwise all of the views have to use `model.get('propertyName')` instead of `model.propertyName`, which is annoying. So, it's turned into JSON first to make it less annoying to use.

File: src/views/components/ListingContent.jsx [hunk 3]
```diff
@@ -215,7 +215,11 @@ class ListingContent extends React.Component {
         }
       }
       if (src) {
-        style.backgroundImage = 'url(' + src + ')';
+        var a = document.createElement('a');
+        a.href = src;
+        a.protocol = location.protocol;
```
Comment (dwick) [inline src/views/components/ListingContent.jsx]: images have a protocol and it's not always https

Comment (dwick) [inline src/views/components/ListingContent.jsx]: can break this into a separate commit if that's wanted

**Comment by dwick** [comment]
:haircut: this is up on staging too https://m.staging.wireddit.com

**Comment by curioussavage** [comment]
:fish:

**Comment by ajacksified** [comment]
:fish:
